src/concurrency.md

```diff
@@ -0,0 +1,760 @@
+changing its electrical / logical state, and the arrival of a new byte of data.
+In some multi-core devices a core can send an interrupt signal to a different
+core.
+
+How the processor locates the right interrupt handler to execute depends on the
+architecture. In the ARM Cortex-M architecture, there's one handler per
+interrupt signal and there's a table somewhere in memory that holds function
+pointers to all interrupt handlers. Each interrupt is given an index in this
+table. For example, a timer interrupt could be interrupt #0 and an input pin
+interrupt could be interrupt #1. If we were to depict this as Rust code it would
+look as follows:
+
+``` rust
+// `link_section` places this in some known memory location
+#[link_section = ".interrupt_table"]
+static INTERRUPT_TABLE: [extern "C" fn(); 32] = [
+    // entry 0: timer 0
+    on_timer0_interrupt,
+
+    // entry 1: pin 0
+    on_pin0_interrupt,
+
+    // .. 30 more entries ..
+];
+
+// provided by the application author
+extern "C" fn on_timer0_interrupt() {
+    // ..
+}
+
+extern "C" fn on_pin0_interrupt() {
+    // ..
+}
+```
+
+In another common interrupt model all interrupts signals map to the *same*
+interrupt handler (subroutine) and there's a hardware register that the software
+has to read when it enters the handler to figure out which interrupt signal
+triggered the interrupt. In this text, we'll focus on the ARM Cortex-M
+architecture which follows the one handler per interrupt signal model.
+
+## Interrupt handling API
+
+The most basic interrupt handling API lets the programmer *statically* register
+a function for each interrupt handler *only once*. On top of this basic API
+it's possible to implement APIs to *dynamically* register closures as interrupt
+handlers. In this text we'll focus on the former, simpler API.
+
+To illustrate this kind of API let's look at the [`cortex-m-rt`] crate (v0.6.7).
+It provides two attributes to statically register interrupts: `#[exception]` and
+`#[interrupt]`. The former is for device agnostic interrupts, whose number and
+names are the same for all Cortex-M devices; the latter is for device specific
+interrupts, whose number and names vary per device / vendor. We'll stick to the
+device agnostic interrupts ("exceptions") in our examples.
+
+[`cortex-m-quickstart`]: https://github.com/rust-embedded/cortex-m-quickstart
+[`cortex-m-rt`]:  https://crates.io/crates/cortex-m-rt/0.6.7
+
+The following example showcases the system timer (`SysTick`) interrupt, which
+fires periodically. The interrupt is handled using the `SysTick` handler
+(function), which prints a dot to the console.
+
+> **NOTE:** The code for the following example and all other examples can be
+> found in the `ci/concurrency` directory at the root of [this repository].
+
+[this repository]: https://github.com/rust-embedded/embedonomicon
+
+``` rust
+{{#include ../ci/concurrency/examples/systick.rs}}```
```

Comment:
The backticks on the same line break with the current mdbook version (there's multiple instances of this).

`````suggestion
{{#include ../ci/concurrency/examples/systick.rs}}
```
`````